### PR TITLE
gh-110395: invalidate open kqueues after fork

### PR DESCRIPTION
**Pull request by sorcio**
Track `select.kqueue` objects with an open fd in a linked list, so they can be invalidated after fork. The invalidation happens through `posix.register_at_fork()` for lack of a better/safer idea.

This PR does some extra allocations, but the number of kqueues you'll open in a program is normally very small, and otherwise it's very lightweight (tracking is only involved at open/close/fork).

I thought of different approaches but I'm not convinced:
- we could check the process pid. But it effectively requires to check at every operation on the kqueue. Performance impact would not be major, but still requires all code to pay for a corner case.
- add some sort of `.detach()` method to invalidate the object without closing the fd. Leaves the burden of correct cleanup to users.
- don't call `close()` on object destructor. Might make sense if combined with a check on pid at creation/destruction time. But still leaves some burden to users.

* Issue: gh-110395

### CODE REVIEW

**Review by gpshead — COMMENTED** [review]
Overall I think this PR makes sense. Additional eyeballs from others such as @vstinner on what looks like module C API modernization with the tp_dealloc removal in favor of a tp_finalize would be helpful.

**Review by gpshead — APPROVED** [review]

**Comment by gpshead** [comment]
I double checked the tp_dealloc and tp_finalize stuff and gave this another look. it's good, thanks!

**Comment by vstinner** [comment]
test_asyncio test_unhandled_exceptions() failed on Windows x86: it's a known bug, fixed in the main branch. I re-run the job. If it fails too often, you can consider to rebase your PR on the main branch to get my test_unhandled_exceptions() fix.
